### PR TITLE
NTC constraints management

### PR DESCRIPTION
**Pull request by Milos-RTEi**
Set NTC to zero or local values depending on the type of the area (adequacy area type) and user defined options

### CODE REVIEW

**Comment by Milos-RTEi** [comment]
@hugo-antoine-rtei did you get the feedback from RTE/Jean-Marc where should we put the UI options for NTC. Optimization preferences or advanced parameters?

**Comment by Milos-RTEi** [comment]
The function "SetNTCForAdequacyFirstStep" somehow does not belong to "opt_gestion_des_bornes_cas_lineaire.cpp". I thought of moving it to "math.h" but only pure math functions are in this file. So still wondering where is the best place. If anybody has a suggestion feel free..

**Comment by Milos-RTEi** [comment]
https://github.com/rte-i/Antares_Simulator/pull/8/commits/7320db6dd89f0311b9113b2584a607b15e588c7b#:~:text=Xmin%5BVar%5D%20%3D%20ValeursDeNTC,ValeursDeNTC%2D%3EValeurDeNTCExtremiteVersOrigine%5BInterco%5D)%3B

This should maybe be discussed with RTE. When managing/altering the NTC constraints first step of adequacy patch, depending on the options, new types of constraint are possible [0 +inf] and [-inf 0]. The second one creates problem to the solver and Antares provides no results. All other constraints [-value1 +value2] (value can be zero) and [-inf +inf] works fine in test cases.

File: src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp [hunk 1]
```diff
@@ -177,10 +230,37 @@ void OPT_InitialiserLesBornesDesVariablesDuProblemeLineaire(PROBLEME_HEBDO* Prob
...
+            {
+                Xmax[Var] = 0.;
+                Xmin[Var] = 0.;
+            }
+            else if (SetToZeroLinkNTCForAdequacyPatchFirstStep == Data::setStartEndToZero)
+            {
+                Xmax[Var] = 0.;
+                Xmin[Var] = ValeursDeNTC->ValeurDeNTCExtremiteVersOrigine[Interco] > ADEQUACY_MAX
+                              ? -ADEQUACY_MAX
+                              : -(ValeursDeNTC->ValeurDeNTCExtremiteVersOrigine[Interco]);
```
Comment (Milos-RTEi) [inline src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]: This should maybe be discussed with RTE. When managing/altering the NTC constraints for the first step of adequacy patch, depending on the options, new types of constraint are possible [0 +inf] and [-inf 0]. The second one creates problem to the solver and Antares provides no results. All other constraints [-value1 +value2] (value can be zero) and [-inf +inf] works fine in test cases.

**Comment by rtei-pw** [comment]
> The function "SetNTCForAdequacyFirstStep" somehow does not belong to "opt_gestion_des_bornes_cas_lineaire.cpp". I thought of moving it to "math.h" but only pure math functions are in this file. So still wondering where is the best place. If anybody has a suggestion feel free..

Create a new file, like adequacy_patch.cpp

File: src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp [hunk 2]
```diff
@@ -41,6 +41,7 @@
 #include "spx_constantes_externes.h"
 
 #define EPSILON_DEFAILLANCE 1e-3
+#define ADEQUACY_MAX 10e+12
```
Comment (rtei-pw) [inline src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]: is there any other MAX value in the projet? this is a quite big value, could be problem for solver

Comment (Milos-RTEi) [reply src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]: Тhis is just a WIP value, I used to make the test cases run properly. __DBL_MAX__ is used as inf. There are some other MAX values for other stuff like max logs, max memory.. but cold not find anything useful for this case.

**Review by rtei-pw — APPROVED** [review]

File: src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp [hunk 3]
```diff
@@ -140,6 +142,12 @@ void OPT_InitialiserLesBornesDesVariablesDuProblemeLineaire(PROBLEME_HEBDO* Prob
     double* Xmin;
     double* Xmax;
     int* TypeDeVariable;
+    bool AdequacyFirstStep;
+    uint StartNodeAdequacyPatchType;
+    uint EndNodeAdequacyPatchType;
+    bool SetToZero12LinksForAdequacyPatch;
+    bool SetToZero11LinksForAdequacyPatch;
```
Comment (hugo-antoine-rtei) [inline src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]: I think that:
``uint StartNodeAdequacyPatchType;``
``uint EndNodeAdequacyPatchType;``
``bool SetToZero12LinksForAdequacyPatch;``
``bool SetToZero11LinksForAdequacyPatch;``
do not need to be explicitly defined

File: src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp [hunk 4]
```diff
@@ -177,10 +189,37 @@ void OPT_InitialiserLesBornesDesVariablesDuProblemeLineaire(PROBLEME_HEBDO* Prob
...
-
-            Xmin[Var] = -(ValeursDeNTC->ValeurDeNTCExtremiteVersOrigine[Interco]);
+            StartNodeAdequacyPatchType = ProblemeHebdo->StartAreaAdequacyPatchType[Interco];
+            EndNodeAdequacyPatchType = ProblemeHebdo->EndAreaAdequacyPatchType[Interco];
+            SetToZeroLinkNTCForAdequacyPatchFirstStep
+              = SetNTCForAdequacyFirstStep(AdequacyFirstStep,
+                                           StartNodeAdequacyPatchType,
+                                           EndNodeAdequacyPatchType,
+                                           SetToZero12LinksForAdequacyPatch,
+                                           SetToZero11LinksForAdequacyPatch);
```
Comment (hugo-antoine-rtei) [inline src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]: If we don't define them explicitly:
```suggestion
            SetToZeroLinkNTCForAdequacyPatchFirstStep
              = SetNTCForAdequacyFirstStep(AdequacyFirstStep,
                                           ProblemeHebdo->StartAreaAdequacyPatchType[Interco],
                                           ProblemeHebdo->EndAreaAdequacyPatchType[Interco],
                                           true,
                                           true);
```

File: src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp (hunk 1 above)
Comment (hugo-antoine-rtei) [reply src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]: The part just after, with the ``Math::Infinite`` is supposed to take care of the infinity issue by changing the variable type (to bounded or not bounded) in order to set the adequate infinite value. The infinite value is going to depend on the solver used, and Antares is supposed to be compatible with at least two solvers. Therefore I think we should keep it to:
``Xmin[Var] = -(ValeursDeNTC->ValeurDeNTCExtremiteVersOrigine[Interco]);``
and report the problem to RTE.

File: src/solver/optimisation/adequacy_patch.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+** SPDX-License-Identifier: licenceRef-GPL3_WITH_RTE-Exceptions
+*/
+#ifndef __SOLVER_ADEQUACY_FUNCTIONS_H__
+#define __SOLVER_ADEQUACY_FUNCTIONS_H__
+
+uint SetNTCForAdequacyFirstStep(bool AdequacyFirstStep,
+                                uint StartNodeAdequacyPatchType,
+                                uint EndNodeAdequacyPatchType,
+                                bool SetToZero12LinksForAdequacyPatch,
+                                bool SetToZero11LinksForAdequacyPatch);
```
Comment (hugo-antoine-rtei) [inline src/solver/optimisation/adequacy_patch.h]: We're missing the documentation of the function, I propose:
```
/**
 * Determines if a link capacity needs to be set to 0. Only changes something if used during the AdequacyFirstStep.
 *
 * @param AdequacyFirstStep boolean for the first run of the optimization used by the adequacy patch
 * 
 * @param StartNodeAdequacyPatchType uint: The adq type of the node at the start of the link.
 * 
 * @param EndNodeAdequacyPatchType uint: The adq type of the node at the end of the link.
 * 
 * @param SetToZero12LinksForAdequacyPatch bool: Switch to cut links from nodes of adq type 1 towards nodes of adq type 2
 * 
 * @param SetToZero11LinksForAdequacyPatch bool: Switch to cut links from nodes of adq type 1 towards nodes of adq type 1
 * 
 * @return uint from an enumeration that describes the type of restrictions to put on this link for adq purposes.
 */
```

File: src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp (hunk 2 above)
Comment (hugo-antoine-rtei) [reply src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]: There is already supposed to be a security built in for it inside the file, with that DBL_MAX and the use of the yuni `Math::Infinite`. So I think that any issue regarding infinity values should be passed on to RTE.

File: src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp [hunk 6]
```diff
@@ -208,13 +247,19 @@ void OPT_InitialiserLesBornesDesVariablesDuProblemeLineaire(PROBLEME_HEBDO* Prob
...
+                {
+                    if (CoutDeTransport->IntercoGereeAvecLoopFlow == OUI_ANTARES)
+                        Xmax[Var] = ValeursDeNTC->ValeurDeNTCOrigineVersExtremite[Interco]
+                                    - ValeursDeNTC->ValeurDeLoopFlowOrigineVersExtremite[Interco];
+                    else
+                        Xmax[Var] = ValeursDeNTC->ValeurDeNTCOrigineVersExtremite[Interco];
 
-                Xmax[Var] += 0.01;
+                    Xmax[Var] += 0.01;
+                }
```
Comment (hugo-antoine-rtei) [inline src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]: This is about transportation cost, I don't think we need to be touching that ?

File: src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp [hunk 7]
```diff
@@ -226,13 +271,20 @@ void OPT_InitialiserLesBornesDesVariablesDuProblemeLineaire(PROBLEME_HEBDO* Prob
...
+                {
+                    if (CoutDeTransport->IntercoGereeAvecLoopFlow == OUI_ANTARES)
+                        Xmax[Var] = ValeursDeNTC->ValeurDeNTCExtremiteVersOrigine[Interco]
+                                    + ValeursDeNTC->ValeurDeLoopFlowOrigineVersExtremite[Interco];
+                    else
+                        Xmax[Var] = ValeursDeNTC->ValeurDeNTCExtremiteVersOrigine[Interco];
 
-                Xmax[Var] += 0.01;
+                    Xmax[Var] += 0.01;
+                }
```
Comment (hugo-antoine-rtei) [inline src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]: same here, at this point, the Var is the transportation cost

**Review by hugo-antoine-rtei — CHANGES_REQUESTED** [review]
Hello,
I left a couple of suggestions, but overall this looks great !

File: src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp (hunk 1 above)
Comment (Milos-RTEi) [reply src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]: OK. Let's report to RTE.

File: src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp (hunk 2 above)
Comment (Milos-RTEi) [reply src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]: OK. Pass to RTE.

File: src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp (hunk 3 above)
Comment (Milos-RTEi) [reply src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]: Done.

File: src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp (hunk 4 above)
Comment (Milos-RTEi) [reply src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]: Done.

File: src/solver/optimisation/adequacy_patch.h (hunk 5 above)
Comment (Milos-RTEi) [reply src/solver/optimisation/adequacy_patch.h]: Done.

File: src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp (hunk 6 above)
Comment (Milos-RTEi) [reply src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]: Yes. I tested and debugged this part. 
It is only activated if Hurdle costs are used and Var is indeed the transportation cost.
I though just to be sure :). But you are most probably right, it is not necessary.  
Ejected.

File: src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp (hunk 7 above)
Comment (Milos-RTEi) [reply src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]: Ejected.

**Comment by hugo-antoine-rtei** [comment]
> @hugo-antoine-rtei did you get the feedback from RTE/Jean-Marc where should we put the UI options for NTC. Optimization preferences or advanced parameters?

Jean-Marc told me it could be in any of the two. I guess since we put the Adequacy patch switch inside the Optimization preferences, it would make sense to have these two other options next to the Adequacy patch box.

**Review by hugo-antoine-rtei — APPROVED** [review]

**Pull request merged**